pandas/core/internals/construction.py

```diff
@@ -595,22 +602,82 @@ def _list_of_dict_to_arrays(data, columns, coerce_float=False, dtype=None):
     data = [(type(d) is dict) and d or dict(d) for d in data]
 
     content = list(lib.dicts_to_array(data, list(columns)).T)
-    return _convert_object_array(
-        content, columns, dtype=dtype, coerce_float=coerce_float
-    )
+    columns = _validate_or_indexify_columns(content, columns)
+    content = _convert_object_array(content, dtype=dtype, coerce_float=coerce_float)
+    return content, columns
 
 
-def _convert_object_array(content, columns, coerce_float=False, dtype=None):
+def _validate_or_indexify_columns(
+    content: List, columns: Union[Index, List, None]
+) -> Union[Index, List[Axis]]:
+    """If columns is None, make numbers as column names; If not None, validate if
+    columns are valid in length.
+
+    Parameters
+    ----------
+    content: list of processed data records
+    columns: Iterable or None
+
+    Returns
+    -------
+    columns: If columns is Iterable, return as is; If columns is None, assign
+    positional column index value as columns.
+
+    Raises
+    ------
+    1. When content is not composed of list of lists, and if length of columns
+        is not equal to length of content.
+    2. When content is list of lists, but length of each sub-list is not equal
+    3. When content is list of lists, but length of sub-list is not equal to
```

Comment:
can we say what kinds of exceptions are raised in these cases?

Reply:
changed!